src/treecstr.ml

```diff
@@ -0,0 +1,341 @@
+(** This module provides a model of binary constraints to be used in decision
+    trees. *)
+open Terms
+open Extra
+
+(** Holds a constraint to solve and its heuristic score, or nothing if there
+    are no constraint available. *)
+type 'a cdecision = ('a * float) option
+
+(** Binary constraints allow to check properties on terms during evaluation.
+    A constraint is binary as it gives birth to two trees, one used if the
+    constraint is satisfied and the other if not.
+
+    Currently, binary constraints are used
+    - to check non linear constraints in the left hand side of a rule (e.g. in
+      presence of the rule like [f &X &X (s &Y) → &Y]).  In this case, the
+      constraint node created will force the rewriting engine to verify that
+      the terms at position [{0}] and [{1}] are convertible.
+    - to verify which variables are free in a term.  If there is a rule of the
+      form [f &X[x, y] &Y[y] → &Y], then the rewriting engine must verify that
```

Comment:
Abstractions are missing and &Y does not have the same arity on both sides. instead, use the rule: f (\x\y,&X[x]) --> r for instance.